browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc

```diff
@@ -52,6 +56,53 @@ constexpr char kURLManagePremium[] = "https://account.brave.com/";
 
 namespace ai_chat {
 
+namespace {
+
+#if !BUILDFLAG(IS_ANDROID)
+// Invokes a callback when the WebContents has finished loading. Note: If the
+// WebContents is destroyed before loading is completed, the callback will not
+// be invoked.
+// The lifetime of this class is tied to the WebContents it is observing - it
+// will be destroyed when |WebContentsDestroyed| is called, or when the
+// Navigation finishes, whichever happens first.
+class WaitForCommit : public content::WebContentsObserver {
+ public:
+  WaitForCommit(
+      content::WebContents* contents,
+      base::OnceCallback<void(content::WebContents* contents)> on_loaded)
+      : WebContentsObserver(contents), on_loaded_(std::move(on_loaded)) {}
+  ~WaitForCommit() override = default;
+
+  void DidFinishNavigation(content::NavigationHandle* handle) override {
+    if (handle->IsInMainFrame() && handle->HasCommitted()) {
+      std::move(on_loaded_).Run(web_contents());
+      delete this;
+    }
+  }
+
+  void WebContentsDestroyed() override { delete this; }
+
+ private:
+  base::OnceCallback<void(content::WebContents* contents)> on_loaded_;
+};
+
+// Note: After session restore we need to ensure the WebContents is loaded
```

Comment:
session restore is not the only time this could happen, tabs can be unloaded for other reasons